### PR TITLE
Convert core components to TypeScript (Round 2)

### PR DESCRIPTION
**Pull request by trblackw**
@nickmcmillan @vaibhavhrt Okay, everything now should be able to be merged smoothly considering index.js (now index.tsx) is hook-based. I apologize for the merge conflict hiccup.

### CODE REVIEW

**Comment by nickmcmillan** [comment]
@trblackw thanks for your work on this (and no stress about the merge conflict)! 

However I'm having a little trouble getting this branch to run. I'm seeing the message `../dist/index.es.js Module not found: Can't resolve './styles.css'`. 

I've run `yarn` to ensure all dependencies are updated. Is there something else I'm missing?

**Comment by trblackw** [comment]
Yeah I attempted to address that my adding the resolve plugin to the roll
up config. TypeScript doesn’t like the import statement when importing
files that aren’t .ts/.tsx/.js. The require syntax that I used at the top
of the refactored index.tsx combined with the plugin added to the roll up
config should have fixed this 🧐 I can look into it

**Comment by trblackw** [comment]
This might help
https://link.medium.com/wE7MP8SVdX

I’m away from my computer for the next day so (vacation from work) if you
could maybe find a solution before then maybe that would be easier?
Otherwise I can address it in the next couple days

**Comment by vaibhavhrt** [comment]
@trblackw waiting for you to complete this before I can start working on #11

**Comment by trblackw** [comment]
I can’t seem to replicate the issue you were having on my end. When are you seeing the error message you listed above? When running `yarn start`? I just pushed a change to include `.css` extensions in the path options within the rollup config. Let me know if this resolves the issue on your end.

**Comment by vaibhavhrt** [comment]
@trblackw you need to merge upstream/master

**Comment by trblackw** [comment]
@nickmcmillan Forgive me because I'm not very familiar with working with forked copies of repos. I'm currently on `master` of my fork and getting this error when attempting to merge:
Screen Shot 2019-06-12 at 6 47 00 AM
I'm assuming there's some precursor steps I need to follow in order to be able to execute that command successfully? 

edit: I did however just `git pull git@github.com:nickmcmillan/react-physics-dragger.git` and then resolve any merge conflicts and push.

**Comment by vaibhavhrt** [comment]
Make sure you are on branch master:

    git checkout master

Fetch remote:

    git fetch

Merge remote:

    git merge upstream/master

**Comment by trblackw** [comment]
@vaibhavhrt What's weird is `git fetch` did not result in the fetching of the origin branch, which is why running `git merge upstream/master` was resulting in the screen shot I added above, I presume. I'm a bit of a git noob as well so I probably messed something up along the way. I see the conflicts have been resolved after what I just pushed, does that mean we're good to go?

I am noticing that Travis CI isn't very happy though
Screen Shot 2019-06-12 at 6 55 35 AM
These environment variables weren't something I messed with on my end

**Comment by vaibhavhrt** [comment]
In conflict resolving you probably forced your changes instead of accepting incoming changes so eslint is giving errors.
Go `src/package.json`, find the `lintJs` commands in scripts section and add fix option to it:

    "lintJs": "eslint **/*.{js,jsx} --fix"

Run it: `npm run lintJs`, it will fix travis errors.
undo changes in package.json, so it should be `"lintJs": "eslint **/*.{js,jsx}"` again.
Also make sure you have deleted `src/index.js`

**Comment by trblackw** [comment]
Okay will have this done by 12:00 EST. Apologies for making this more
difficult than it should be.

**Comment by vaibhavhrt** [comment]
Actually forget my last comment, you actually messed up eslint settings, nothing wrong in resolving conflicts. We were using 2 spaces for tabs, you changed it 4 and trailing comma was allowed, you removed it too. See my review for fixing it.

File: .eslintrc [hunk 1]
```diff
@@ -1,26 +1,32 @@
...
-  },
-  "rules": {
-    // don't force es6 functions to include space before paren
-    "space-before-function-paren": 0,
-
-    // allow specifying true explicitly for boolean props
-    "react/jsx-boolean-value": 0,
-
-    // allow comma-dangle in multiline objects & arrays
-    "comma-dangle": ["error", "only-multiline"]
```
Comment (vaibhavhrt) [inline .eslintrc:24]: This line needs to be there, add it back.

    "comma-dangle": ["error", "only-multiline"]

**Comment by trblackw** [comment]
👍🏼

File: .eslintrc [hunk 2]
```diff
@@ -1,26 +1,32 @@
...
-  }
-}
+   "parser": "@typescript-eslint/parser",
+   "extends": [
+     "standard",
+     "standard-react",
+     "plugin:@typescript-eslint/recommended"
```
Comment (vaibhavhrt) [inline .eslintrc]: Extending this rule is probably forcing 4 spaces for indent. Add this line in rules section to make tab size 2 spaces:

    "indent": ["error", 2],

See https://stackoverflow.com/a/48906878/9321755 for more details.

**Review by vaibhavhrt — CHANGES_REQUESTED** [review]
Fix these two 👇 , run `npm run lintJs` to make sure there are no more errors, if there is fix them.

**Comment by trblackw** [comment]
Okay I won’t be able to push until 11:00 EST so I’ll let you know when I do

**Comment by vaibhavhrt** [comment]
K, let me know if there are still any errors

**Comment by trblackw** [comment]
@vaibhavhrt okay I updated the eslint settings and fixed the affected files. I also just removed the tslint plugin because it was directly conflicting with eslint.

File: src/.eslintrc [hunk 3]
```diff
@@ -1,5 +0,0 @@
-{
-  "env": {
-    "jest": true
```
Comment (vaibhavhrt) [inline src/.eslintrc:3]: put this line in actual `.eslintrc`.

**Review by vaibhavhrt — CHANGES_REQUESTED** [review]
@trblackw add `rollup.config.js` in `.eslintignore` to ignore it(or fix the errors if you want).

File: package.json [hunk 4]
```diff
@@ -51,12 +53,19 @@
...
   "files": [
     "dist"
-  ]
-}
\ No newline at end of file
+  ],
+  "dependencies": {
+    "@types/jest": "^24.0.13",
+    "@types/react": "^16.8.19",
+    "@types/react-dom": "^16.8.4"
```
Comment (vaibhavhrt) [inline package.json]: `@types` dependencies are devDependencies as well. Move them to `devDependencies`

**Review by vaibhavhrt — CHANGES_REQUESTED** [review]
Also add `react-scripts-ts` as devDependency and then update the test scripts(in package.json) to run ts tests, right now it's not finding any tests as it's looking for tests with extension `js`.

    "test": "cross-env CI=1 react-scripts-ts test --env=jsdom",
    "test:watch": "react-scripts-ts test --env=jsdom",

**Comment by trblackw** [comment]
Okay. I’m offline again for a bit so I’ll address this this afternoon

**Comment by trblackw** [comment]
@vaibhavhrt let me know if there's anything else I need to do

**Comment by vaibhavhrt** [comment]
> Also add `react-scripts-ts` as devDependency and then update the test scripts(in package.json) to run ts tests, right now it's not finding any tests as it's looking for tests with extension `js`.
> 
> ```
> "test": "cross-env CI=1 react-scripts-ts test --env=jsdom",
> "test:watch": "react-scripts-ts test --env=jsdom",
> ```

**Comment by trblackw** [comment]
Oops, forgot that one. Should be good now @vaibhavhrt

**Comment by vaibhavhrt** [comment]
Ok looks like ts requires a tsconfig.test.json to be able to run test suite.
Add a `tsconfig.test.json` in root directory with following contents.

    {
      "extends": "./tsconfig.json",
      "compilerOptions": {
        "module": "commonjs"
      }
    }

**Comment by trblackw** [comment]
@vaibhavhrt done

**Comment by vaibhavhrt** [comment]
@trblackw test is now running which is good, but it's failing. I am not sure what's wrong, might need some investigation. Can you try to figure out what's wrong. You can run tests locally by `npm run test` to make sure it passes before pushing commits.
You can see error details by running tests locally or see the logs on tavis.

**Comment by trblackw** [comment]
I’ll take a look when I get a chance

**Comment by trblackw** [comment]
@vaibhavhrt So I'm been trying to trouble shoot what's going on with the test that's failing in `index.tsx`, but I'm not super familiar with testing (sad face). Perhaps you might have a bit more insight into what's potentially going wrong here
Screen Shot 2019-06-16 at 9 00 54 AM

What's confusing to me is that the error message: `Unexpected token (221:12)` is actually on line 248 in my editor (perhaps this is normal because extra spaces are removed upon compilation or something?). My editor shows no visible issues with the code. The result message of the test is `Test suite failed to run`, which seems to contradict the previous unexpected token issue?

What do you think?

**Comment by vaibhavhrt** [comment]
@trblackw I will clone your fork of this repo and take a look later. rn busy watching world cup

File: src/index.tsx [hunk 5]
```diff
@@ -1,53 +1,70 @@
-import React, { useEffect, useRef, useState } from 'react'
-import PropTypes from 'prop-types'
+// eslint-disable-next-line
+import React, { useEffect, useRef, useState, MouseEvent, RefObject, ReactNode } from 'react'
 
 import { roundNum } from './utils'
 import { applyDragForce, applyBoundForce } from './force'
 import getBoundaries from './getBoundaries'
+// TypeScript does not like using the import statement for non js/ts files
+const styles = require('./styles.css')
```
Comment (vaibhavhrt) [inline src/index.tsx:8]: Actually you can import css files, you just need to create a `typings.d.ts` in src with following content:

    /**
     * Default CSS definition for typescript,
     * will be overridden with file-specific definitions by rollup
     */
    declare module "*.css" {
      const content: { [className: string]: string };
      export default content;
    }

File: src/index.tsx [hunk 6]
```diff
@@ -228,35 +249,24 @@ export default function Dragger(props) {
...
       ref={outerEl}
-      className={`${styles.outer} ${isDraggingStyle ? styles.isDragging : ''}${props.disabled ? ' is-disabled' : ''} ${props.className}`}
+      className={`${styles.outer} ${isDraggingStyle ? styles.isDragging : ''}${props.disabled
+        ? ' is-disabled'
+        : ''} ${props.className}`}
       onTouchStart={onStart}
       onMouseDown={onStart}
-      style={{ ...props.style }}
-    >
+      style={{ ...props.style }}>
```
Comment (vaibhavhrt) [inline src/index.tsx]: Why spread the `props.style` and create a new object, you can just do `style={props.style}`
```suggestion
      style={props.style}>
```

File: tsconfig.json [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+     "resolveJsonModule": true,
+     "isolatedModules": true,
+     "downlevelIteration": true,
+     "noEmit": true,
+     "jsx": "preserve"
+   },
+   "include": [
+     "src"
+   ]
+ }
```
Comment (vaibhavhrt) [inline tsconfig.json:24]: There is something wrong with `tsconfig.json`, I am not expert enough to figure out what exactly. But here is the config I use. I tested it on clone of your fork, tests are passing after updating tsconfig.

    {
      "compilerOptions": {
        "outDir": "build",
        "module": "esnext",
        "target": "es5",
        "lib": ["es6", "dom", "es2016", "es2017"],
        "sourceMap": true,
        "allowJs": false,
        "jsx": "react",
        "declaration": true,
        "moduleResolution": "node",
        "forceConsistentCasingInFileNames": true,
        "noImplicitReturns": true,
        "noImplicitThis": true,
        "noImplicitAny": true,
        "strictNullChecks": true,
        "suppressImplicitAnyIndexErrors": true,
        "noUnusedLocals": true,
        "noUnusedParameters": true
      },
      "include": ["src"],
      "exclude": ["node_modules", "build", "dist", "example", "rollup.config.js"]
    }

Maybe you will know what exactly is causing the error and update those parts only. Or just copy paste this config, it works pretty well for me.

**Review by vaibhavhrt — CHANGES_REQUESTED** [review]
Problem is with tsconfig

File: src/index.tsx (hunk 6 above)
Comment (trblackw) [reply src/index.tsx]: I didn't write the personally.

File: src/index.tsx (hunk 5 above)
Comment (trblackw) [reply src/index.tsx:8]: this didn't work for me. I'm leaving the rollup config modifications for now. test is now passing

**Comment by trblackw** [comment]
@vaibhavhrt Okay after using the `tsconfig.json` you provided above the test is passing

File: src/index.tsx (hunk 5 above)
Comment (vaibhavhrt) [reply src/index.tsx:8]: yeah no need to actually do this. I am just telling its possible.

**Comment by vaibhavhrt** [comment]
Looks good to me now. @nickmcmillan you can review and merge.

**Review by vaibhavhrt — APPROVED** [review]

**Comment by nickmcmillan** [comment]
I'm still unable to get the example to run in this branch. Steps to reproduce are;

1. Checkout this branch, run `yarn` to ensure dependencies are updated
2. `yarn start` in the root folder
3. `cd example` and `yarn start` to launch the development example

At this point I see the error message I'd mentioned here; https://github.com/nickmcmillan/react-physics-dragger/pull/9#issuecomment-498341603 - which appears both in the terminal and in the browser.

4. I implemented the solution @vaibhavhrt suggested here; https://github.com/nickmcmillan/react-physics-dragger/pull/9#discussion_r294463338 which worked - I'm no longer seeing the error message.

In the browser, I'm seeing a new error message (screenshot attached) which suggests to me that Typescript is not being compiled, the browser is attempting to read TS directly.

Screenshot 2019-06-18 at 13 25 59


Am I missing something? Have you both tried testing the project example as per step #3 above and if so did it work?

I've noticed that the tool used to bootstrap this project (https://github.com/transitive-bullshit/create-react-library) also has a Typescript scaffolding option. Is this something worth considering as a last resort? (ie re-scaffolding the project using create-react-library with TS enabled)

**Comment by vaibhavhrt** [comment]
@trblackw you can take a look at this repo, I am using ts there, it works (almost) perfectly, [vaibhavhrt/react-github-buttons](https://github.com/vaibhavhrt/react-github-buttons). My project was also bootstrapped using `create-react-library` but I used the typescript template, and manually updated all dev-dependencies.

**Comment by trblackw** [comment]
@vaibhavhrt Why not just use `create-react-app` with the `--typescript` flag? I've worked with TypeScript a lot, mostly with the foundation set up with CRA and have never had issues. @nickmcmillan I'm curious why you used `create-react-library`?

**Comment by vaibhavhrt** [comment]
@trblackw because it's a npm library not a react website/app. This repo is supposed to be pluggable into some other react app(may or may not be bootstrapped with CRA).
When creating an app we have an `index.html` but not in a library like this, because it will be used in some app which will have it `index.html`.
Web app (usually) have minified js, css etc, here in library we don't minify it. So that users can easily debug, and minify, gzip according to their choice.
In library we publish (usually) `index.js` and `index.es.js` to enable tree shaking.

**Comment by trblackw** [comment]
Oh, well that makes sense then. Very cool. That’s a bit out of my scope,
though, as far as how to go about resolving the current compile issue.

**Comment by vaibhavhrt** [comment]
I am guessing something is wrong in rollup.config Take a look at the rollup config of my repo and try to see what's different in them. I am not expert enough in ts to figure out the solution just by looking at error, so you try changing things in rollup.config If you can't figure it out I will clone your fork again and take a look.

**Comment by trblackw** [comment]
Okay I’ll take a look. Will get back to you some time tomorrow @vaibhavhrt

**Comment by trblackw** [comment]
@vaibhavhrt Haven't had any luck. Maybe you could take a look

**Comment by vaibhavhrt** [comment]
ok, I will look when i get some time

**Comment by nickmcmillan** [comment]
I think I have a working solution. @vaibhavhrt and @trblackw would you mind taking a look at this PR; https://github.com/nickmcmillan/react-physics-dragger/pull/12 (which is branched of the work you've both done so far in this current PR)

Basically I re-initialised the `create-react-library` with TS enabled which provided a working foundation. I then copied the `src` files over from this PR, fixed a few errors, and it seems to be up and running ok (using the `/example` project).

@vaibhavhrt I don't think we need `lintJs` anymore and the eslint plugins that go with it. We can just use `create-react-library`'s testing set up. So I've just made `lintJS` an alias for `test` so that Travis tests can pass.

**Pull request merged**